### DISCUSSION TITLE
Reraise exception in yield except block but get FastAPIError finally.

### DISCUSSION THREAD

**Discussion by seokhyuncheong1** [fastapi/fastapi#11832] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
async def http_exception_handler(request: Request, exc: HTTPException):
    base_error_message = f'http exception: {request.method}: {request.url} with {request.path_params}'
    logger.error(f'{traceback.format_exc()}')

    return JSONResponse(
        status_code=exc.status_code,
        content=set_default_response_content(
            CommonCode.EXC_HTTP, base_error_message, exc.detail)
    )


async def redis_session() -> AsyncGenerator[Redis, None]:
    async with RedisPool() as client:
        try:
            yield client
        except Exception as e:
            logger.warning(f'redis_session error: {traceback.format_exc()}')
            raise e


app = FastAPI(
    title=f'{app_config.APP_NAME}@{app_config.VERSION}',
    exception_handlers={
        HTTPException: http_exception_handler,
        ...
        # other exceptions
    }
)
        

@app.get('/api/v1/test')
async def test_router(test: Optional[str] = None, redis: Redis = Depends(redis_session)):
    if test is None:
        raise HTTPException(status_code=400, detail='test is required')
    
    return {'test': test}
```

### Description

Hi, our teams are considering updating FastAPI to v0.111.0.
After update FastAPI to v0.111.0, we suffer due to FastAPIError.
we use yield object in logic, and reraise exception in except block like example code.

We use redis session with `Depends()`.
When we raise exception, fastapi return FastAPIError response and we can see below error msg.
```
No response object was returned. There's a high chance that the application code is raising an exception and a dependency with yield has a block with a bare except, or a block with except Exception, and is not raising the exception again. Read more about it in the docs:
https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#dependencies-with-yield-and-except
```

We add exception handler to FastAPI and it is not working...
Also, we return JSONResponse in exception handler.
Please, give me some solutions...

### Operating System

Linux

### FastAPI Version

0.111.0

### Pydantic Version

2.6.1

### Python Version

3.11.8

**Comment by YuriiMotov** [comment]
Can't reproduce your issue.
Could you provide a code example that we can copy\paste, run and see the error?

**Reply by seokhyuncheong1** [reply]
Here is sample code to run!
You can see the error in log (not in response).
```py
import logging
from typing import Optional
from fastapi import (
    FastAPI,
    HTTPException,
    Request,
    Depends
)
from fastapi.responses import JSONResponse


logger = logging.getLogger(__name__)


class TestRedisPool:
    _client = None

    async def __aenter__(self):
        self._client = []
        return self._client

    async def __aexit__(self, exc_type, exc_value, tb):
        return self


async def http_exception_handler(request: Request, exc: HTTPException):
    base_error_message = f'http exception: {request.method}: {request.url} with {request.path_params}'

    return JSONResponse(
        status_code=exc.status_code,
        content={
            'message': base_error_message,
            'detail': exc.detail
        }
    )


async def redis_session():
    async with TestRedisPool() as client:
        try:
            yield client
        except Exception as e:
            logger.error(f'Failed to connect redis: {e}')
            raise e


app = FastAPI(
    title='test_app',
    exception_handlers={
        HTTPException: http_exception_handler
    }
)


@app.get('/api/v1/test')
async def test_router(test: Optional[str] = None, redis=Depends(redis_session)):
    if test is None:
        raise HTTPException(status_code=400, detail='test is required')

    return {'test': test}


if __name__ == '__main__':
    import uvicorn
    uvicorn.run(
        'test_app:app',
        host='0.0.0.0',
        port=5050,
        reload=True,
        workers=1,
        log_level='info',
    )
```

**Reply by YuriiMotov** [reply, marked as answer]
The problem here is that you return `self` from `__aexit__`.
`with` statement re-raises exceptions only if the return value of `__aexit__` is `False`:

> If the suite was exited due to an exception, and the return value from the [__exit__()](https://docs.python.org/3/reference/datamodel.html#object.__exit__) method was false, the exception is reraised. If the return value was true, the exception is suppressed, and execution continues with the statement following the [with](https://docs.python.org/3/reference/compound_stmts.html#with) statement.

https://docs.python.org/3/reference/compound_stmts.html#with


So, just change it to 
```
    async def __aexit__(self, exc_type, exc_value, tb):
        return False
```
and everything works

**Reply by seokhyuncheong1** [reply]
Thank you so so so so so much!!!
Everything works fine!!
